File: 백준/Silver/6603. 로또/로또.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")
  .map((nums) => nums.split(" ").map(Number));

const winNumbers = [];

for (let i = 0; i < input.length; i++) {
  if (input[i][0] === 0) {
    break;
  }
  input[i].shift();
  const sortedArr = input[i].sort((a, b) => a - b);
  dfs(0, 0, sortedArr);
  if (i !== input.length - 2) {
    console.log();
  }
}

function dfs(depth, start, arr) {
  if (depth === 6) {
    console.log(winNumbers.join(" "));
    return;
  }

  for (let i = start; i < arr.length; i++) {
    if (!winNumbers.includes(arr[i])) {
      winNumbers.push(arr[i]);
      start++;
      dfs(depth + 1, start, arr);
      winNumbers.pop(arr[i]);
    }
  }
}

File: 백준/Silver/6603. 로또/로또.py
while True:
    nums = list(map(int, input().split()))
    if len(nums) == 1 and nums[0] == 0:
        break

    k = nums.pop(0)

    stack = []
    visited = [0] * k

    def dfs(depth, start):
        if depth == 6:
            print(*stack)
            return

        for i in range(start, k):
            if not visited[i]:
                visited[i] = 1
                stack.append(nums[i])
                dfs(depth + 1, i)
                stack.pop()
                visited[i] = 0

    dfs(0, 0)
    print('')
